File: Python/Algorithms/RadixSort.py
def FindMaxValue(elements):
    maxval = elements[0]
    for val in elements:
        if val>maxval:
            maxval = val
    return maxval


def radixSort(myarr):
    base = 10
    temparr = [0 for i in range(base)]
    copy = [0 for i in range(len(myarr))]

    # Finding the maxvalue by max function or by hard coding
    # maxval = max(myarr)
        # or 
    maxval = FindMaxValue(myarr)

    position = 1
    while(maxval//position>0):
        # make temparr 0
        for i in range(10): temparr[i] = 0

        # Perform Count Sort

        for val in myarr:
            idx = (val//position)%base
            temparr[idx]+=1
        
        for i in range(1,10):
            temparr[i]+=temparr[i-1]

        for i in range(9,0,-1):
            temparr[i] = temparr[i-1]
        temparr[0] =0 
        
        for val in myarr:
            idx = (val//position)%base 
            copy[temparr[idx]] = val 
            temparr[idx]+=1
        
        for i in range(len(myarr)):
            myarr[i] = copy[i]

        position = position*10;

# Stable Algorithm

arr = [24,56,89,35,129,422]
radixSort(arr)

print(*arr)

File: C++/Algorithms/RadixSort.cpp
#include<iostream>
using namespace std;

// Stable Algorithm

int FindMaxValue(int data[] , int size){
    int maxval = data[0];
    for(int i = 1 ;i<size;i++){
        if (data[i]>maxval){
            maxval = data[i];
        }
    }
    return maxval;
}

void radixSort(int myarr[] , int size){
    int base = 10;
    int temparr[base];
    int copy[size];
    int maxval = FindMaxValue(myarr,size);

    int position = 1;

    while((maxval/position)>0){
        // Filling 0 to temparr
        for(int i =0;i<10;i++) temparr[i] = 0;
        
        // Perform Count sort

        for(int i = 0;i<size;i++){
            int idx = (myarr[i]/position)%base;
            temparr[idx]++;
        }
        
        for(int i = 1;i<10;i++){
            temparr[i]+=temparr[i-1];
        }

        for(int i = 9 ; i>0;i--){
            temparr[i] = temparr[i-1];
        }
        temparr[0] = 0;


        for(int i =0;i<size;i++){
            int idx = (myarr[i]/position)%base;
            copy[temparr[idx]] = myarr[i];
            temparr[idx]++;
        }

        for(int i =0;i<size;i++){
            myarr[i] = copy[i];
        }
        position*=10;
    }
}

int main(){
    int arr[] = {24,56,89,35,129,422};
    int size = sizeof(arr)/sizeof(arr[0]);

    radixSort(arr,size);

    for(int i = 0;i<size;i++){
        cout<<arr[i]<<" ";
    }

    return 0;
}